File: src/main/java/com/mikea/bayes/FactorProduct.java
package com.mikea.bayes;

/**
 * @author mike.aizatsky@gmail.com
 */
//todo FactorProduct should have the same interface as Factor.
public class FactorProduct {
    private final Factor[] factors;

    public FactorProduct(Factor[] factors) {
        this.factors = factors;
    }

    public Factor compute() {
        Factor result = factors[0];
        for (int i = 1; i < factors.length; i++) {
            result = result.product(factors[i]);
        }
        return result;
    }

    public double computeAt(VarAssignment assignment) {
        double result = 1;

        for (Factor factor : factors) {
            double value = factor.getValue(assignment);
            result *= value;
        }

        return result;
    }
}

File: src/main/java/com/mikea/bayes/FactorProduct.scala
package com.mikea.bayes

class FactorProduct(val factors: Seq[Factor]) {
  def compute(): Factor = {
    var result = factors(0)
    for (i <- 1 until factors.length) {
      result = result.product(factors(i))
    }
    result
  }

  def computeAt(assignment: VarAssignment): Double = {
    var result = 1.0
    for (factor <- factors) {
      val value = factor.getValue(assignment)
      result *= value
    }
    result
  }
}
